src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1461,6 +1507,65 @@ private RefreshItem getRefreshToken(final AuthenticationRequest request) {
         return refreshItem;
     }
 
+    private TokenCacheItem findFamilyItemForUser(final UserIdentifierType userIdentifierType, final String userId)
+    {
+        final String methodName = ":findFamilyItemForUser";
+            
+        if (StringExtensions.IsNullOrBlank(userId))
+        {
+            Logger.v(TAG + methodName, "User id is not provided, cannot continue with finding family item.");
+            return null;
+        }
+
+        Iterator<TokenCacheItem> allItems = mTokenCacheStore.getAll();
+        if (allItems == null || !allItems.hasNext())
+        {
+            Logger.v(TAG + methodName, "No items in the cache, cannot continue with finding family item.");
+        }
+        
+        if (userIdentifierType == UserIdentifierType.UniqueId)
+        {
+            Logger.v(TAG + methodName, "UserIdentifier type is unique id, will look for the family item based on unique id.");
+            while (allItems.hasNext())
+            {
+                final TokenCacheItem tokenCacheItem = allItems.next();
+                final UserInfo tokenUserInfo = tokenCacheItem.getUserInfo();
+                if (tokenUserInfo!= null 
+                    && tokenUserInfo.getUserId() != null 
+                    && tokenUserInfo.getUserId().equalsIgnoreCase(userId))
+                {
+                    if (!StringExtensions.IsNullOrBlank(tokenCacheItem.getFamilyClientId())
```

Comment:
Then you don't need familyId as part of the key, just use username. It'll simplify the logic a lot.
I don't get why do you need to store this flag in Cache, because if you store it by different CachKey it will be obvious that you put it there because of this flag.
And then within getRefreshToken you'll need something like 
            if (item == null || StringExtensions.IsNullOrBlank(item.getRefreshToken())) {
                // if not present, check universal item for family
                Logger.v(TAG, "Looking for Refresh token with Family Client Id");
                keyUsed = userId;
                item = mTokenCacheStore.getItem(keyUsed);
                familyResource = true;
            }
                refreshItem = new RefreshItem(keyUsed, request, item, multiResource, familyResource);
It will simplify logic a lot